video: gc2145: addition of CSI output support

avolmat-st opened the pull request:
The PR adds GC2145 driver support to allow controlling a CSI based GC2145 sensor.
This is the setup used on the STM32MP13-DK board in which case the GC2145 is connected to the ST-MIPID02 CSI bridge (PR to be pushed as well) into the DCMIPP (ongoing PR 89407 with added parallel support to be pushed).

The current driver currently support 320x240 / 640x480 and 1600x1200 resolution, however for the time being I haven't been able to make the 320x240 resolution work in CSI (probably a PIXELRATE issue with the CSI receiver timing to be adjusted ...) so I only allow 640x480 and 1600x1200 resolution in CSI for the moment.

I've also added standard CSI2 DT macros into video.h so that not all sensor or receiver will have to define them.

I do not have a setup with a DVP GC2145 so could someone run a sanity test with those patches in a DVP mode.

Commit:
video: gc2145: add CSI mode of the GC2145 sensor

Add possibility to use the gc2145 sensor in CSI mode by
adding the bus-type property in the device-tree.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: gc2145: avoid default DVP enabling

Do not enable by default the DVP streaming to avoid conflict
with the introduction of the MIPI mode of the GC2145.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: addition of CSI2 standard data-type defines

Add standard data-type macros within the video.h in order
to avoid having to redefine them all in each and every CSI
based driver.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: gc2145: use VGA resolution as default

Since QVGA resolution is currently not working in CSI, use VGA
as default resolution to avoid failing during the init of the
driver.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented on dts/bindings/video/galaxycore,gc2145.yaml:
```diff
@@ -18,3 +18,7 @@ properties:
     description: |
       The PWDN pin is asserted to power down the sensor. The sensor
       receives this as an active high signal
+
+child-binding:
+  child-binding:
+    include: video-interfaces.yaml
```
Thank you for helping with #80514 in the process!

The Arducam Nicla Vision board, only one using the GC2145 in boards/, uses `remote-endpoint` as a placeholder. Enabling `video-interfaces.yaml` prevent it to build:

`devicetree error: 'remote-endpoint-label' is marked as required in 'properties:' in /home/tinyvision/zephyrproject/zephyr/dts/bindings/video/galaxycore,gc2145.yaml, but does not appear in <Node /soc/i2c@40005c00/gc2145@3c/port/endpoint in '/home/tinyvision/zephyrproject/zephyr/misc/empty_file.c'>`

`FATAL ERROR: command exited with status 1: /bin/cmake -DWEST_PYTHON=/home/tinyvision/.local/bin/python3 -B/home/tinyvision/zephyrproject/zephyr/build -GNinja -DBOARD=arduino_nicla_vision/stm32h747xx/m7 -S/home/tinyvision/zephyrproject/zephyr/samples/subsys/usb/uvc`

This fixes it:

```diff
patch -p1 <<EOF
diff --git a/boards/arduino/nicla_vision/arduino_nicla_vision_stm32h747xx_m7.dts b/boards/arduino/nicla_vision/arduino_nicla_vision_stm32h747xx_m7.dts
index 0028128f8e2..669a2266825 100644
--- a/boards/arduino/nicla_vision/arduino_nicla_vision_stm32h747xx_m7.dts
+++ b/boards/arduino/nicla_vision/arduino_nicla_vision_stm32h747xx_m7.dts
@@ -225,7 +225,7 @@ zephyr_udc0: &usbotg_hs {
 
 		port {
 			gc2145_ep_out: endpoint {
-				remote-endpoint = <&dcmi_ep_in>;
+				remote-endpoint-label = "dcmi_ep_in";
 			};
 		};
 
@@ -252,7 +252,7 @@ zephyr_udc0: &usbotg_hs {
 
 	port {
 		dcmi_ep_in: endpoint {
-			remote-endpoint = <&gc2145_ep_out>;
+			remote-endpoint-label = "gc2145_ep_out";
 		};
 	};
 };
EOF
```

avolmat-st replied:
Actually this fix is already part of the PR #89627 so better not fix it in this PR as well. Any apparently PR #89627 is already partially approved.

josuah replied:
Thanks that makes sense!

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -1433,6 +1433,29 @@ static inline unsigned int video_bits_per_pixel(uint32_t pixfmt)
+ *
+ * @{
+ */
+#define VIDEO_MIPI_CSI2_DT_YUV422_8	0x1e
+#define VIDEO_MIPI_CSI2_DT_YUV422_10	0x1f
+#define VIDEO_MIPI_CSI2_DT_RGB444	0x20
+#define VIDEO_MIPI_CSI2_DT_RGB555	0x21
+#define VIDEO_MIPI_CSI2_DT_RGB565	0x22
+#define VIDEO_MIPI_CSI2_DT_RGB666	0x23
+#define VIDEO_MIPI_CSI2_DT_RGB888	0x24
+#define VIDEO_MIPI_CSI2_DT_RAW6		0x28
+#define VIDEO_MIPI_CSI2_DT_RAW7		0x29
+#define VIDEO_MIPI_CSI2_DT_RAW8		0x2a
+#define VIDEO_MIPI_CSI2_DT_RAW10	0x2b
+#define VIDEO_MIPI_CSI2_DT_RAW12	0x2c
+#define VIDEO_MIPI_CSI2_DT_RAW14	0x2d
```
This seems to mostly match the naming of [`include/media/mipi-csi2.h`](https://elixir.bootlin.com/linux/v6.11.1/source/include/media/mipi-csi2.h).

josuah reviewed (commented):
Using the [patch soup](https://github.com/zephyrproject-rtos/zephyr/discussions/89691) snippets to test with UVC:

```
git remote add tvai https://github.com/tinyvision-ai-inc/zephyr
git fetch tvai
git cherry-pick tvai/pr-usb-uvc~9..tvai/pr-usb-uvc
west build -b arduino_nicla_vision/stm32h747xx/m7 samples/subsys/usb/uvc/
west flash
```

![mpv-shot0002](https://github.com/user-attachments/assets/e4744a6b-d4e3-405a-803c-a848190e4c4e)

This still works!

josuah reviewed (changes requested):
Suggesting a change for the small syntax bugfix `remote-endpoint` -> `remote-endpoint-label` if that is ok to ask.
Otherwise I can open a separate PR as it is not a bug introduced by this and ~~CI does not fail due to this~~.

[EDIT: It does but https://github.com/zephyrproject-rtos/zephyr/pull/89627 is where the fix is]

avolmat-st commented:
Corrected the PIXEL_RATE part that I got wrong within the sensor side when doing the ST_MIPID02 driver at the same time.
With this capture is fine on STM32MP13xx (DCMIPP) + ST_MIPID02 + GC2145.
I'm also going to push the ST_MIPID02 PR now.

avolmat-st commented:
Small update to avoid if (err) style error checking and instead use MISRA validated if (err < 0).

josuah commented on drivers/video/gc2145.c:
```diff
@@ -42,6 +42,54 @@ LOG_MODULE_REGISTER(video_gc2145, CONFIG_VIDEO_LOG_LEVEL);
+#define GC2145_DPHY_MODE2_CLK_DIFF(a)		((a) & 0x07)
+#define GC2145_DPHY_MODE2_LANE0_DIFF(a)	(((a) & 0x07) << 4)
+
+#define GC2145_REG_DPHY_MODE3	0x03
+#define GC2145_DPHY_MODE3_LANE1_DIFF(a)	((a) & 0x07)
+#define GC2145_DPHY_MODE3_CLK_DELAY		BIT(4)
+#define GC2145_DPHY_MODE3_LANE0_DELAY		BIT(5)
+#define GC2145_DPHY_MODE3_LANE1_DELAY		BIT(6)
+
+#define GC2145_REG_FIFO_FULL_LVL_LOW	0x04
+#define GC2145_REG_FIFO_FULL_LVL_HIGH	0x05
+#define GC2145_REG_FIFO_MODE		0x06
+#define GC2145_FIFO_MODE_READ_GATE	BIT(3)
+#define GC2145_FIFO_MODE_MIPI_CLK_MODULE	BIT(7)
+
+#define GC2145_REG_BUF_CSI2_MODE	0x10
+#define GC2145_CSI2_MODE_DOUBLE		BIT(0)
+#define GC2145_CSI2_MODE_RAW8		BIT(2)
+#define GC2145_CSI2_MODE_MIPI_EN	BIT(4)
+#define GC2145_CSI2_MODE_EN		BIT(7)
+
+#define GC2145_REG_MIPI_DT		0x11
+#define GC2145_REG_LWC_LOW		0x12
+#define GC2145_REG_LWC_HIGH		0x13
+#define GC2145_REG_DPHY_MODE		0x15
+#define GC2145_DPHY_MODE_TRIGGER_PROG	BIT(4)
+
+#define GC2145_REG_FIFO_GATE_MODE	0x17
+#define GC2145_REG_T_LPX		0x21
+#define GC2145_REG_T_CLK_HS_PREPARE	0x22
+#define GC2145_REG_T_CLK_ZERO		0x23
+#define GC2145_REG_T_CLK_PRE		0x24
+#define GC2145_REG_T_CLK_POST		0x25
+#define GC2145_REG_T_CLK_TRAIL		0x26
+#define GC2145_REG_T_HS_EXIT		0x27
+#define GC2145_REG_T_WAKEUP		0x28
+#define GC2145_REG_T_HS_PREPARE		0x29
+#define GC2145_REG_T_HS_ZERO		0x2a
+#define GC2145_REG_T_HS_TRAIL		0x2b
+
```
I will try to port it to the CCI library, I [got started for DVP](https://github.com/zephyrproject-rtos/zephyr/commit/7b9d6af7d26bcc4930fa2521be232b399d42c512) and will do it after the merge

josuah commented on drivers/video/gc2145.c:
```diff
@@ -1205,6 +1403,8 @@ static const struct gc2145_config gc2145_cfg_0 = {
 #if DT_INST_NODE_HAS_PROP(0, reset_gpios)
 	.reset_gpio = GPIO_DT_SPEC_INST_GET(0, reset_gpios),
 #endif
+	.bus_type = DT_PROP_OR(DT_CHILD(DT_INST_CHILD(0, port), endpoint), bus_type,
+			       VIDEO_BUS_TYPE_PARALLEL),
```
It might be possible to use the macro `DT_INST_ENDPOINT_BY_ID(inst, pid, eid) for walking down the port and endpoint`:

https://github.com/zephyrproject-rtos/zephyr/blob/c35bb0de8023750de5ba31dcfbd7b5277466147e/include/zephyr/devicetree/port-endpoint.h#L105-L163

josuah reviewed (commented):
If interested in improving this now, using `DT_INST_ENDPOINT_BY_ID(inst, pid, eid)` can be nicer, but I do not see this as blocking at least IMHO.

Thank you for extending this sensorr!

avolmat-st commented:
I fixed the macro to get the bus-type. Should have done that already in the first place. Thanks for noticing.

avolmat-st commented:
Apparently some tests are not successful but I have to say that I do not understand what is failing.

josuah commented:
> Apparently some tests are not successful but I have to say that I do not understand what is failing.

Maybe the CI issues will get resolved once https://github.com/zephyrproject-rtos/zephyr/pull/89571/files#r2080630046 is resolved (fix: merge 

> `devicetree error: 'remote-endpoint-label' is marked as required in 'properties:' in .../galaxycore,gc2145.yaml, but does not appear in <Node /soc/i2c@40005c00/gc2145@3c/port/endpoint>`
> https://github.com/zephyrproject-rtos/zephyr/actions/runs/14953906764/job/42007171754?pr=89571#step:12:927

avolmat-st commented:
> > Apparently some tests are not successful but I have to say that I do not understand what is failing.
> 
> Maybe the CI issues will get resolved once https://github.com/zephyrproject-rtos/zephyr/pull/89571/files#r2080630046 is resolved (fix: merge #89627 ?)
> 

oh yeah indeed. That #89627 should now be ready I think so we should be able to merge it before this one. Or maybe adding in the description that this PR depends on the other one might help the CI build ?

josuah commented:
> Or maybe adding in the description that this PR depends on the other one might help the CI build ?

You might need to rebase this PR on top of #89627's commits. Alas Github is not able to grok the order in which it should apply patches like maintainers would in a mailing list.

Commit:
video: add LINK_FREQ ctrl and a helper to retrieve it

Add a ctrl VIDEO_CID_LINK_FREQ to indicate to a sink
the frequency at which a device streams data over
CSI2. Since not all source device currently provide the
LINK_FREQ control, add a helper function to retrieve
it, with a fall-back by doing an approximate via the
PIXEL_RATE control.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: common: addition of INTEGER_MENU control type

Add a new INTEGER_MENU type allowing to store signed 64bits
integer into a menu.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

loicpoulain commented on include/zephyr/drivers/video.h:
```diff
@@ -1433,6 +1433,29 @@ static inline unsigned int video_bits_per_pixel(uint32_t pixfmt)
+#define VIDEO_MIPI_CSI2_DT_RAW10	0x2b
+#define VIDEO_MIPI_CSI2_DT_RAW12	0x2c
+#define VIDEO_MIPI_CSI2_DT_RAW14	0x2d
+
```
Would be good to define all the CSI2 data types now?

avolmat-st replied:
Alright yes !! Just converting to draft for the moment in order to avoid polluting the PR listing, while updating this.

avolmat-st commented:
The MIPI data type macros are removed since I moved them into ST_MIPID02 PR which is a dependency for this PR since it also introduce the LINK_FREQUENCY ctrl. Last push allows you to see the latest modification (introduction of LINK_FREQUENCY). Then I will rebase this PR once ST_MIPID02 serie get agreed and CCI PR get merged.
(this PR will have to be rebased on top of main branch to also take into account the removal of the endpoint parameters).

josuah commented on drivers/video/gc2145.c:
```diff
@@ -1082,6 +1236,14 @@ static int gc2145_set_fmt(const struct device *dev, enum video_endpoint_id ep,
 	}
 
+	if (cfg->bus_type == VIDEO_BUS_TYPE_CSI2_DPHY) {
+		ret = gc2145_config_csi(dev, fmt->pixelformat, fmt->width, fmt->height);
```
If the format is not supported for CSI but it is for DVP (like 320x240), then `drv_data->fmt = *fmt` would still be applied, but the format would not be valid. Probably not a big issue though.

avolmat-st replied:
Moved the drv_data->fmt setting after CSI check.

josuah reviewed (dismissed):
We might need to review it again when dependencies get merged, but LGTM in the meantime.
One minor issue observed... that is probably me chasing ghosts rather than finding bugs!

avolmat-st commented:
Simply rebased on HEAD.

josuah reviewed (dismissed):
![mpv-shot0002](https://github.com/user-attachments/assets/072cc3a4-86c0-4802-ab4c-af1474181ab1)

Still works!

I could not spot any issue left. LGTM thank you!

avolmat-st commented:
Following a comment in PR 89764 which contains the 
video: common: addition of INTEGER_MENU control type
video: add LINK_FREQ ctrl and a helper to retrieve it

I update this PR to also have the same valid commits.

Now that LINK_FREQ ctrl is no more READ_ONLY by default, add setting the READ_ONLY flag within the driver here since with this driver the LINK_FREQ ctrl is readl-only.
